### PR TITLE
Admin delete content

### PR DESCRIPTION
**Pull request by janderson16**
Adds functionality for an admin to delete a listing
-adds migration to delete cascading (images)
-adds admin listings controller (show, index, delete)
-adds admin_listings_index and admin_listing_show views w/ delete buttons
-adds a sad path to messaging to ensure users cannot view each others messages

99 examples, 0 failures, 1 pending

### CODE REVIEW

File: app/controllers/admin/listings_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  def show
+    @listing = Listing.find(params[:id])
+  end
+
+  def destroy
+    listing = Listing.find(params[:id])
+    listing.images.delete
+    listing.delete
+    flash[:notice] = "#{listing.id} successfully deleted"
+    redirect_to admin_listings_path
```
Comment (meyerhoferc) [inline app/controllers/admin/listings_controller.rb:16]: 👍

File: app/models/user.rb [hunk 2]
```diff
@@ -6,7 +6,7 @@ class User < ApplicationRecord
   enum status: [:active, :inactive]
 
 
-  has_many :listings
+  has_many :listings, dependent: :destroy
```
Comment (meyerhoferc) [inline app/models/user.rb]: 👍

File: app/services/permission.rb [hunk 3]
```diff
@@ -59,5 +59,6 @@ def admin_user_permissions
     return true if controller == "admin/users"
     return true if controller == "sessions"
     return true if controller == "listings"
+    return true if controller == "admin/listings"
```
Comment (meyerhoferc) [inline app/services/permission.rb:62]: 👍

File: app/views/admin/listings/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                              <td><%= link_to listing.title, admin_listing_path(listing) %></td>
+                              <td><%= listing.street_address %>
+                                  <%= listing.city %>
+                                  <%= listing.state %>
+                                  <%= listing.zipcode %></td>
+                              <td><%= listing.list_category %></td>
+                              <td><%= listing.max_occupancy %> person(s)</td>
+                              <td>$<%= listing.cost_per_night %></td>
+                              <td><%= button_to "Delete", admin_listing_path(listing), method: :delete, class: "btn btn-success" %></td>
+                            </div>
```
Comment (meyerhoferc) [inline app/views/admin/listings/index.html.erb:43]: great use of a table here

File: config/routes.rb [hunk 5]
```diff
@@ -17,6 +17,7 @@
   namespace :admin do
     get 'dashboard', to: "dashboard#show"
     resources :users, only: [:index, :update]
+    resources :listings, only: [:index, :show, :destroy]
```
Comment (meyerhoferc) [inline config/routes.rb:20]: 👍

File: db/migrate/20170403023324_add_cascade_delete_to_images.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+class AddCascadeDeleteToImages < ActiveRecord::Migration[5.0]
+  def change
+    remove_foreign_key :images, :listings
+    add_foreign_key :images, :listings, on_delete: :cascade
```
Comment (meyerhoferc) [inline db/migrate/20170403023324_add_cascade_delete_to_images.rb:4]: interesting, I've never seen this on a migration before

File: spec/features/admin/admin_can_delete_listing_spec.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    admin.roles.create(title: "admin")
+    allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(admin)
+
+    visit admin_listing_path(listing)
+
+    within(".admin") do
+      click_button("Delete")
+    end
+    expect(current_path).to eq(admin_listings_path)
+    expect(page).to_not have_content("Big House")
```
Comment (meyerhoferc) [inline spec/features/admin/admin_can_delete_listing_spec.rb:42]: great tests

File: spec/features/messages/user_can_send_message_to_another_user_spec.rb [hunk 8]
```diff
@@ -31,4 +31,20 @@
...
+    host = Fabricate(:user)
+    host.roles.create!(title: "host")
+    allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(user_2)
+
+    conversation_1 = Conversation.create!(initiator_id: host.id, recipient_id: user_1.id)
+    message_1 = conversation_1.messages.create!(user_id: user_1.id, body: "msg test 1")
+
+    visit "/conversations/#{conversation_1.id}/messages"
+
+    expect(page).to have_content("404")
```
Comment (meyerhoferc) [inline spec/features/messages/user_can_send_message_to_another_user_spec.rb:47]: great addition

**Review by meyerhoferc — APPROVED** [review]
looks great, merging now

File: db/migrate/20170403023324_add_cascade_delete_to_images.rb (hunk 6 above)
Comment (janderson16) [reply db/migrate/20170403023324_add_cascade_delete_to_images.rb:4]: @meyerhoferc ya had to do some digging, for whatever reason 'dependent :destroy' wasn't doing the trick

**Pull request merged**